api/src/pcapi/core/offers/api.py

```diff
@@ -305,9 +307,27 @@ def update_collective_offer(
             raise educational_exceptions.OffererOfVenueDontMatchOfferer()
 
     nationalProgramId = new_values.pop("nationalProgramId", None)
-    national_program_api.link_or_unlink_offer_to_program(nationalProgramId, offer_to_update)
+    try:
+        national_program_api.link_or_unlink_offer_to_program(nationalProgramId, offer_to_update, commit=False)
+    except Exception:
+        db.session.rollback()
+        raise
 
-    updated_fields = _update_collective_offer(offer=offer_to_update, new_values=new_values)
+    with atomic():
+        updated_fields = _update_collective_offer(offer=offer_to_update, new_values=new_values, commit=False)
+
+        try:
```

Comment:
J'aurais bien fait en sorte que la condition `if new_venue` englobe le traitement des réservations. Ça me semble plus logique, car la valeur de `new_venue` ne dépend pas des réservations sur lesquelles on itère.

Au final (avec mon commentaire ci-dessous), j'ai l'impression qu'on pourrait écrire ça comme suit : 

```python
updated_fields = _update_collective_offer(offer=offer_to_update, new_values=new_values, commit=False)
if new_venue:
    for booking in offer_to_update.collectiveStock.collectiveBookings:
        booking.venue = new_venue
```

Reply:
En effet, et on peut donc passer complètement via un `.update()`. Confusion de ma part en `offer.venue` et `offer.offerVenue`.